modules/core/src/main/java/org/apache/ignite/internal/managers/communication/IgniteMessageFactoryImpl.java

```diff
@@ -0,0 +1,166 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.managers.communication;
+
+import java.lang.reflect.Array;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Supplier;
+
+import org.apache.ignite.IgniteException;
+import org.apache.ignite.plugin.extensions.communication.IgniteMessageFactory;
+import org.apache.ignite.plugin.extensions.communication.Message;
+import org.apache.ignite.plugin.extensions.communication.MessageFactory;
+import org.apache.ignite.plugin.extensions.communication.MessageFactoryProvider;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.annotations.TestOnly;
+
+/**
+ * Message factory implementation which is responsible for instantiation of all communication messages.
+ */
+public class IgniteMessageFactoryImpl implements IgniteMessageFactory {
+    /** Offset. */
+    private static final int OFF = -Short.MIN_VALUE;
+
+    /** Array size. */
+    private static final int ARR_SIZE = 1 << Short.SIZE;
+
+    /** Custom messages registry. Used for test purposes. */
+    private static final Map<Short, Supplier<Message>> CUSTOM = new ConcurrentHashMap<>();
+
+    /** Message suppliers. */
+    private final Supplier<Message>[] msgSuppliers = (Supplier<Message>[]) Array.newInstance(Supplier.class, ARR_SIZE);
+
+    /** Initialized flag. If {@code true} then new message type couldn't be registered. */
+    private boolean initialized;
+
+    /**
+     * Contructor.
+     *
+     * @param factories Concrete message factories or message factory providers. Cfn't be empty or {@code null}.
+     */
+    public IgniteMessageFactoryImpl(MessageFactory[] factories) {
+        if (factories == null || factories.length == 0)
+            throw new IllegalArgumentException("Message factory couldn't be initialized. Factories aren't provided.");
+
+        List<MessageFactory> old = new ArrayList<>(factories.length);
+
+        for (MessageFactory factory : factories) {
+            if (factory instanceof MessageFactoryProvider) {
+                MessageFactoryProvider p = (MessageFactoryProvider)factory;
+
+                p.registerAll(this);
+            }
+            else
+                old.add(factory);
+        }
+
+        if (!old.isEmpty()) {
+            for (int i = 0; i < ARR_SIZE; i++) {
+                Supplier<Message> curr = msgSuppliers[i];
+
+                if (curr == null) {
+                    short directType = indexToDirectType(i);
+
+                    for (MessageFactory factory : old) {
+                        Message msg = factory.create(directType);
```

Comment:
Should you wrap this into try-catch?
Specification isn't very clear - it says that method returns null for unresolved direct type but at the same time this class (IgniteMessageFactoryImpl) throws exceptions for that same reason. We should probably unify this behavior and correct tests correspondingly.

Reply:
I agree with you, but:
- at present we have several usages of `msgFactory.create(type)` where `null` is valid but treated in different way (e.g. `DirectByteBufferStreamImplV2#readMessage`);
- exception will thrown to higher level and in current implementation and architecture it is ok.

In the future (possibly during 3.0 development) we should return to this issue.